In [2]:
import numpy as np
import pandas as pd
import ast
# Import linear_kernel
import nltk
import string
import re
import unidecode
#nltk.download('wordnet')
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet
from collections import Counter
import config 
from sklearn.metrics.pairwise import linear_kernel
#Import TfIdfVectorizer (scikit-learn)
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# Loading data and looking at a  sample

In [4]:
#df_82k = pd.read_csv ("C:/Users/Atif-Foxfire/Downloads/RAW_recipes.csv/recipes_82k.csv", sep="," )
df_82k = pd.read_csv ("../data/salad-recipe-data/recipes_82k.csv", sep="," )

In [5]:
df_82k.head(10)

category                                     cooking_method       cuisine  \
0      NaN  ['Set the racks in the middle and upper thirds...  ['American']   
1      NaN  ['Place the eggs in the air fryer basket and c...  ['American']   
2      NaN  ['Air Fryer directions:', 'Preheat air fryer t...  ['American']   
3      NaN  ['Preheat the oven to 400F.', 'Pour 2 tablespo...  ['American']   
4      NaN  ['Cook potatoes in a large pot of salted water...  ['American']   
5      NaN  ['To Make the soup:', 'Heat 1 tablespoon of oi...  ['American']   
6      NaN  ['Preheat the to 350F degrees. Spray a pie dis...  ['American']   
7      NaN  ['Combine sour cream and brown sugar and mix w...  ['American']   
8      NaN  ['Preheat the oven to 450F. Place a silicone l...  ['American']   
9      NaN                                                 []           NaN   

                                               image  \
0  https://www.skinnytaste.com/wp-content/uploads...   
1  https://www.skinnytaste.com/wp-content/uploads...   
2  https://www.skinnytaste.com/wp-content/uploads...   
3  https://www.skinnytaste.com/wp-content/uploads...   
4  https://www.skinnytaste.com/wp-content/uploads...   
5  https://www.skinnytaste.com/wp-content/uploads...   
6  https://www.skinnytaste.com/wp-content/uploads...   
7  https://www.skinnytaste.com/wp-content/uploads...   
8  https://www.skinnytaste.com/wp-content/uploads...   
9  https://www.skinnytaste.com/wp-content/uploads...   

                                         ingredients   prep_time  \
0  ['1 tablespoons extra virgin olive oil', '1 cu...  20 minutes   
1  ['4 large eggs', 'Salt (black pepper, everythi...  15 minutes   
2  ['olive oil spray', '4 about 5 ounce each salm...   5 minutes   
3  ['1/2 cup freshly grated Parmesan (not pre-gra...  15 minutes   
4  ['3 1/2 pounds new potatoes (about 10 peeled a...  10 minutes   
5  ['2 tablespoons high-heat oil', '1 medium yell...   5 minutes   
6  ['cooking spray', '1 3/4 cups diced ham steak ...  15 minutes   
7  ['16 oz strawberries (washed and cut)', '4 oz ...   5 minutes   
8  ['1 cup 5 oz all purpose or white whole wheat ...  15 minutes   
9                                                 []   0 minutes   

                                     recipe_name       serves  \
0                         Mediterranean Sea Bass   4 servings   
1                     Air Fryer Hard Boiled Eggs       4 eggs   
2                Air Fryer Basil-Parmesan Salmon   4 servings   
3                     Everything Parmesan Crisps   4 servings   
4                    Potato and Green Bean Salad  12 servings   
5   Green Detox Soup with Toasted Hemp Gremolata   2 servings   
6                 Ham and Swiss Crustless Quiche   6 servings   
7                          Strawberries Romanoff   5 Servings   
8              Smoked Salmon Breakfast Flatbread   4 servings   
9  Skinnytaste Air Fryer Cookbook: Get a Free 39            0   

                                                tags  
0  Dairy Free, Gluten Free, Keto Recipes, Kid Fri...  
1  Air Fryer Recipes, Dairy Free, Gluten Free, Ke...  
2  Air Fryer Recipes, Gluten Free, Keto Recipes, ...  
3  Gluten Free, Keto Recipes, Kid Friendly, Low C...  
4  Dairy Free, Gluten Free, Kid Friendly, Vegetar...  
5  Dairy Free, Gluten Free, Under 30 Minutes, Veg...  
6  Gluten Free, Keto Recipes, Kid Friendly, Low C...  
7  Gluten Free, Kid Friendly, Low Carb, Under 30 ...  
8   Air Fryer Recipes, Gluten Free, Under 30 Minutes  
9                                  Air Fryer Recipes

In [6]:
df_82k.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 82245 entries, 0 to 82244
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   category        18107 non-null  object
 1   cooking_method  82245 non-null  object
 2   cuisine         82230 non-null  object
 3   image           82245 non-null  object
 4   ingredients     82245 non-null  object
 5   prep_time       58630 non-null  object
 6   recipe_name     82245 non-null  object
 7   serves          82244 non-null  object
 8   tags            82011 non-null  object
dtypes: object(9)
memory usage: 5.6+ MB


## Checking for duplicate recipe and ingrdients

In [7]:
print('No of Unique recipes: ',len(df_82k['recipe_name'].unique().tolist()))
print('No of Unique Ingredients: ', len(df_82k['ingredients'].unique().tolist()))

No of Unique recipes:  55838
No of Unique Ingredients:  61198


## Observations
- Dataset consist of 82245 rows and 9 Columns
- Text data no numerical data
- preptime and serves can be converted into numerical data
- a lot of duplicate recipes which can be removed
- some or the rows have 'ingreidentS' value '[]' whch is null but can't be de

In [8]:
# Removing duplicate recipes
df_82k = df_82k.drop_duplicates(subset=['recipe_name'], keep='first')
df_82k = df_82k[df_82k.ingredients != "[]"]
df_82k.reset_index(drop=True, inplace=True)

## Checking for Null / Missing  Values

In [9]:
type(df_82k['ingredients'])
total = df_82k.isnull().sum().sort_values(ascending=False)
percent = (df_82k.isnull().sum()/df_82k.isnull().count()).sort_values(ascending=False)
missing_data = pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])
missing_data.head()

Total   Percent
category   47498  0.863239
prep_time  16466  0.299257
tags         178  0.003235
cuisine        8  0.000145
serves         1  0.000018

## Observations
- A lot of values are missing from 'category' it can be removed

In [10]:
 # dropping category  column
df_82k.drop('category',1, inplace =True)

/tmp/ipykernel_7140/2523923695.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  df_82k.drop('category',1, inplace =True)


# Data Cleaning
## Checking most frequently occuring words in 'ingredients'

In [11]:
vocabulary = nltk.FreqDist()
# This was done once I had already preprocessed the ingredients
for ingredients in df_82k['ingredients']:
    ingredients = ingredients.split()
    vocabulary.update(ingredients)
for word, frequency in vocabulary.most_common(200):
    print(f'{word};{frequency}')

'1;171451
cup;118273
'2;96269
tablespoons;63076
'1/2;60583
teaspoon;59321
and;59274
cups;41954
'1/4;40415
ground;36763
tablespoon;35952
fresh;33730
'3;29706
for;29247
pepper',;28915
or;27776
1/2;27091
'4;26186
oil',;25672
chopped;25267
black;25021
salt',;23817
freshly;22855
chopped',;21589
ounces;20985
olive;20020
sugar',;19997
large;19694
teaspoons;19051
to;18643
into;17573
red;17153
cut;16481
['1;16029
finely;15663
plus;14806
pound;13599
'Kosher;12852
white;12147
of;12062
unsalted;11686
1;11534
butter',;11003
salt;10885
small;10545
'Salt;10476
sliced;10409
garlic,;10405
powder',;10372
['2;10343
flour',;10293
'6;10219
'3/4;9917
grated;9911
sliced',;9788
chicken;9502
onion,;9387
cloves;9308
butter,;9226
'8;9144
juice',;8817
pounds;8699
(about;8693
all-purpose;8651
2;8524
vinegar',;8413
thinly;8233
cream',;8058
medium;8034
peeled;8014
leaves',;7979
vanilla;7962
sauce',;7847
recipe;7743
dried;7330
extra-virgin;7319
'1/3;7224
green;7203
lemon;7079
whole;7038
kosher;7013
water',;6979
diced

## Parsing 'ingredients' column and extractiting Vegetarian recipie and no of servings

In [12]:
def ingredient_parser(ingreds):
    '''
    
    '''
    ingred_list = []
    measure_units = ['teaspoon', 't', 'tsp.', 'tsp','tablespoon', 'T', 'tbl.', 'tb', 'tbsp.', 'fluid ounce', 'fl oz', 'gill', 'cup', 'c', 'pint', 'p', 'pt', 'fl pt', 'quart', 'q', 'qt', 'fl qt', 'gallon', 'g', 'gal', 'ml', 'milliliter', 'millilitre', 'cc', 'mL', 'l', 'liter', 'litre', 'L', 'dl', 'deciliter', 'decilitre', 'dL', 'bulb', 'level', 'heaped', 'rounded', 'whole', 'pinch', 'medium', 'slice', 'pound', 'lb', '#', 'ounce', 'oz', 'mg', 'milligram', 'milligramme', 'g', 'gram', 'gramme', 'kg', 'kilogram', 'kilogramme', 'x', 'of', 'mm', 'millimetre', 'millimeter', 'cm', 'centimeter', 'centimetre', 'm', 'meter', 'metre', 'inch', 'in', 'milli', 'centi', 'deci', 'hecto', 'kilo']
    words_to_remove = ['spicy','allspice','sweetened','bittersweet','golden','flavored','use','coloring','intant','sifted','ice','pure','at','follows','like','sir','granulated','powdered','topping','squeezed','pan','dice','cracked','nonstick','bag','semisweet','plum','whipped','smashed','very','zested','drizzling','well','sharp','reserved','bought','box','cored','salad','reduced','mini','by','julienned','everything','about','each','chopped','diced','together','fresh', 'oil', 'a', 'red', 'bunch', 'and', 'clove', 'or', 'leaf', 'chilli', 'large', 'extra', 'sprig',
    'ground', 'handful', 'free', 'small','brushing','de','slated','recipe','serving','minced', 'pepper', 'virgin', 'range', 'from', 'dried', 'sustainable', 'black', 'peeled', 'higher',
                       'welfare', 'seed', 'for', 'finely', 'freshly', 'sea', 'quality', 'white', 'ripe', 'few', 'piece', 'source', 'to',
                       'organic', 'flat', 'smoked', 'ginger', 'sliced', 'green', 'picked', 'the', 'stick', 'plain', 'plus', 'mixed', 'mint',
                       'bay', 'basil', 'your', 'cumin', 'optional', 'fennel', 'serve', 'mustard', 'unsalted', 'baby', 'paprika', 'fat', 'ask',
                       'natural', 'skin', 'roughly', 'into', 'such', 'cut', 'good', 'brown', 'grated', 'trimmed', 'oregano', 'powder', 'yellow',
                       'dusting', 'knob', 'frozen', 'on', 'deseeded', 'low', 'runny', 'balsamic', 'cooked', 'streaky', 'nutmeg', 'sage', 'rasher',
                       'zest', 'pin', 'groundnut', 'breadcrumb','long-grain', 'turmeric', 'halved', 'grating', 'stalk', 'light', 'tinned', 'dry', 'soft', 'rocket',
                       'bone', 'colour', 'washed', 'skinless', 'leftover', 'splash', 'removed', 'dijon', 'thick', 'big', 'hot', 'drained', 'sized',
                       'chestnut', 'watercress', 'fishmonger', 'english', 'dill', 'caper', 'raw', 'worcestershire', 'flake', 'cider', 'cayenne',
                       'tbsp', 'leg', 'pine', 'wild', 'if', 'fine', 'herb', 'almond', 'shoulder', 'cube', 'dressing', 'with', 'chunk', 'spice', 'thumb',
                       'garam', 'new', 'little', 'punnet', 'peppercorn', 'shelled', 'saffron', 'other','chopped', 'salt', 'olive', 'taste', 'can', 'sauce',
                       'water', 'diced', 'package', 'italian', 'shredded', 'divided', 'parsley', 'vinegar', 'all', 'purpose', 'crushed', 'juice', 'more',
                       'coriander', 'bell', 'needed', 'thinly', 'boneless', 'half', 'thyme', 'cubed', 'cinnamon', 'cilantro', 'jar', 'seasoning', 'rosemary',
                       'extract', 'sweet', 'baking', 'beaten', 'heavy', 'seeded', 'tin', 'vanilla', 'uncooked', 'crumb', 'style', 'thin', 'nut', 'coarsely',
                       'spring', 'chili', 'cornstarch', 'strip', 'cardamom', 'rinsed', 'honey', 'cherry', 'root', 'quartered', 'head', 'softened', 'container',
                       'crumbled', 'frying', 'lean', 'cooking', 'roasted', 'warm', 'whipping', 'thawed', 'corn', 'pitted', 'sun', 'kosher', 'bite', 'toasted', 'lasagna',
                       'split', 'melted', 'degree', 'lengthwise', 'romano', 'packed', 'pod', 'anchovy', 'rom', 'prepared', 'juiced', 'fluid', 'floret', 'room', 'active',
                       'seasoned', 'mix', 'deveined', 'lightly', 'anise', 'thai', 'size', 'unsweetened', 'torn', 'wedge', 'sour', 'basmati', 'marinara', 'dark',
                       'temperature', 'garnish', 'bouillon', 'loaf', 'shell', 'reggiano', 'canola', 'parmigiano', 'round', 'canned', 'ghee', 'crust', 'long',
                       'broken', 'ketchup', 'bulk', 'cleaned', 'condensed', 'sherry', 'provolone', 'cold', 'soda', 'cottage', 'spray', 'tamarind', 'pecorino',
                       'shortening', 'part', 'bottle', 'sodium', 'cocoa', 'grain', 'french', 'roast', 'stem', 'link', 'firm', 'asafoetida', 'mild', 'dash', 'boiling','one','two','three','four']
    
    #Check input ingredient is a list or not if not then convert it into a list. We use ast.literal_eval
    if isinstance(ingreds, list):
        ingredients = ingreds
    else:
        try:
            ingredients = ast.literal_eval(ingreds)
        except:
            return  'none'

    # We first get rid of all the punctuation. We make use of str.maketrans. It takes three input 
    # arguments 'x', 'y', 'z'. 'x' and 'y' must be equal-length strings and characters in 'x'
    # are replaced by characters in 'y'. 'z' is a string (string.punctuation here) where each character
    #  in the string is mapped to None. 
    translator = str.maketrans('', '', string.punctuation)
    lemmatizer = WordNetLemmatizer()
    try:
        for i in ingredients:
            i.translate(translator)
            # We split up with hyphens as well as spaces
            items = re.split(' |-|,', i)
            # Get rid of words containing non alphabet letters
            items = [word for word in items if word.isalpha()]
            # Turn everything to lowercase
            items = [word.lower() for word in items]
            # remove accents
            items = [unidecode.unidecode(word) for word in items] 
            # Lemmatize words so we can compare words to measuring words
            items = [lemmatizer.lemmatize(word) for word in items]
            # Gets rid of measuring units/words
            items = [word for word in items if word not in measure_units]
            # Get rid of commonly occuring words
            items = [word for word in items if word not in words_to_remove]
            if items:
                ingred_list.append(' '.join(items)) 
        ingred_list = " ".join(ingred_list)
    except:
            return ingred_list
    return ingred_list

def vegeterian_parser(ingreds):
    '''
    
    '''
    ingred_list =[]
    non_veg = ['chicken','duck','turkey','buffalo','egg', 'eggs', 'beef','ham','bacon','salami','sausage','pepperoni','steak', 'pork','lamb', 'fish', 'shrimp', 'meat', 'flesh','sardine','tuna','salmon','tilapia','cod',
              'snapper','herring','anchovies','haddock','flounder','trout','catfish','pollock','bass','halibut','sowrdfish','pike',
              'mackerel']
    
    
    #Check input ingredient is a list or not if not then convert it into a list. We use ast.literal_eval
    if isinstance(ingreds, list):
        ingredients = ingreds
    else:
        try:
            ingredients = ast.literal_eval(ingreds)
        except:
            return 'none'

    
    translator = str.maketrans('', '', string.punctuation)
    veg = "yes"
    for i in ingredients:
        i.translate(translator)
        # We split up with hyphens as well as spaces
        items = re.split(' |-|,', i)
        # Get rid of words containing non alphabet letters
        items = [word for word in items if word.isalpha()]
        # Turn everything to lowercase
        items = [word.lower() for word in items]
        # remove accents
        items = [unidecode.unidecode(word) for word in items] 
                
        if items:
            ingred_list.append(' '.join(items)) 
    ingred_list = " ".join(ingred_list)
    ingred_list=  list(ingred_list.split(" "))
    if any(x in non_veg for x in ingred_list):
        veg = "no"
    ''''for word in ingred_list:
        print (word)
        if word  in non_veg:
            veg = "no"'''
    return veg

def serving_extractor(servings):
    translator = str.maketrans('', '', string.punctuation)
    try:
        servings = str(servings)
        servings = re.split(" |-", servings)
        no_of_servings = 0
        for i in range (len(servings)):
            servings[i] = servings[i].translate(translator)
            if servings[i].isdigit():
                no_of_servings = servings[i]
                break
    except:
        return 0
    return no_of_servings
    

In [13]:
#Creating a new column of parsed ingredients vegetarian and no_of_servings
df_82k['parsed_ingredients'] = df_82k['ingredients'].apply(lambda x: ingredient_parser(x))

df_82k['vegetarian'] = df_82k['ingredients'].apply(lambda x: vegeterian_parser(x))

df_82k['no_of_servings'] = df_82k['serves'].apply(lambda x: serving_extractor(x))

#df_82k['parsed_cusine'] = df_82k['cuisine'].apply(lambda x: ingredient_parser(x))


In [14]:
# Dropping Cuisine Preptime and serves columns
df_82k.drop(['cuisine','prep_time','serves'],1, inplace =True)

/tmp/ipykernel_7140/2164615743.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  df_82k.drop(['cuisine','prep_time','serves'],1, inplace =True)


## Looking at new 'parsed_ingredients'

In [15]:
df_82k.sample(10)

cooking_method  \
45524  ['Prepare the ribs: Place the ribs bone-side u...   
33105  ['Heat some canola oil in a pan over medium (n...   
27447  ['Directions', '1.Mix together the herbs, garl...   
26181  ['For the mustard butter: With a rubber spatul...   
46992  ['Fill up a large bowl of water and a few hand...   
35524  ['Place milk, Carnation&reg; Breakfast Essenti...   
52137  ['Preheat the oven to 400 degrees F. Arrange t...   
23716  ['In a medium bowl stir together all ingredien...   
20377  ['Heat a wide, heavy pot over medium high heat...   
14098  ['Directions', 'Working in a small bowl and us...   

                                                   image  \
45524  https://food.fnr.sndimg.com/content/dam/images...   
33105  https://food.fnr.sndimg.com/content/dam/images...   
27447  https://food.fnr.sndimg.com/content/dam/images...   
26181  https://food.fnr.sndimg.com/content/dam/images...   
46992  https://food.fnr.sndimg.com/content/dam/images...   
35524  https://food.fnr.sndimg.com/content/dam/images...   
52137  https://food.fnr.sndimg.com/content/dam/images...   
23716  https://food.fnr.sndimg.com/content/dam/images...   
20377  https://food.fnr.sndimg.com/content/dam/images...   
14098  https://food.fnr.sndimg.com/content/dam/images...   

                                             ingredients  \
45524  ['2 racks St. Louis-style spareribs (3 to 3 1/...   
33105  ['Canola oil, for frying', '6 eggs', 'Worceste...   
27447  ['TENDERLOINS:', '2 turkey tenderloins', '12 c...   
26181  ['1 stick (8 tablespoons) unsalted butter, at ...   
46992  ['Kosher salt and freshly ground black pepper'...   
35524  ['1 packet Classic French Vanilla Flavor Carna...   
52137  ["12 slices thick-cut bacon, such as Nodine's ...   
23716  ['1/3 cup canola oil, plus oil for grate', '2 ...   
20377  ['3 tablespoons extra-virgin olive oil, 3 turn...   
14098  ['1 cup sour cream (you can use nonfat)', '1 t...   

                                             recipe_name  \
45524  St. Louis-Style Ribs with Soy-Ginger Barbecue ...   
33105                       Perfect Fried Egg - Tutorial   
27447  Herb Marinated Turkey Tenderloins with Avocado...   
26181        Grilled Seafood Platter with Mustard Butter   
46992                      String Bean "Casserole" Salad   
35524                                  Pumpkin Pie Shake   
52137                            Ultimate Grilled Cheese   
23716                Garlic, Lemon, and Herb Rubbed Legs   
20377                       Fool-i-ya-baise Seafood Stew   
14098                        Cinnamon Sour Cream Recipe    

                                                    tags  \
45524    Ribs Recipes,Whiskey Recipes,Main Dish,Grilling   
33105                    Egg Recipes,Low-Carb,Low Sodium   
27447  Dip,Avocado,Vegetable,Poultry,Turkey Recipes,M...   
26181  Lobster,Shellfish Recipes,Fish,Salmon,Main Dis...   
46992  Easy Side Dish Recipes,Easy,Side Dish,Green Be...   
35524          Dairy Recipes,Pumpkin,Vegetable,Breakfast   
52137  Grilled Cheese Recipes,Cheese,Bacon Recipes,Me...   
23716  Easy Chicken,Chicken,Easy,Poultry,Easy Main Di...   
20377  French Recipes,Stew Recipes,Fruit,Orange Recip...   
14098  Easy Dessert Recipes,Dessert,Easy,Easy Baking,...   

                                      parsed_ingredients vegetarian  \
45524  rack louis sparerib scotch whiskey bourbon sug...        yes   
33105                                                egg         no   
27447  turkey tenderloin garlic lemon avocado avocado...         no   
26181    butter lemon salmon fillet lobster tail lobster         no   
46992  haricot verts third apple porcini mushroom gar...        yes   
35524  packet classic flavor carnation breakfast comp...        yes   
52137  bacon applewood mayonnaise parmesan cheese pul...         no   
23716         grate grill garlic lemon chicken drumstick         no   
20377  turn palm garlic leek then across then onion r...         no   
14098                      

## Checking most frequently occuring words in 'parsed_ingredients'
(These observation are iterated over many steps to further enhanced the list of words to remove as most commonly occuring words will induce a bias in model's calculatin )

In [16]:
vocabulary = nltk.FreqDist()
# This was done once I had already preprocessed the ingredients
for ingredients in df_82k['parsed_ingredients']:
    ingredients = ingredients.split()
    vocabulary.update(ingredients)
for word, frequency in vocabulary.most_common(200):
    print(f'{word};{frequency}')

sugar;26367
butter;23202
garlic;22097
onion;19182
egg;16605
cream;14510
flour;14407
lemon;13350
cheese;11887
tomato;11791
chicken;11281
milk;8749
wine;8407
vegetable;8027
lime;6705
chocolate;6643
orange;6262
potato;5591
bread;4936
rice;4832
stock;4641
chile;4618
bean;4564
carrot;4433
apple;4075
parmesan;4016
celery;4013
scallion;3838
shallot;3737
beef;3649
mushroom;3372
broth;3346
pork;2957
mayonnaise;2940
bacon;2928
soy;2921
syrup;2882
peanut;2834
jalapeno;2722
breast;2618
paste;2582
chive;2581
chip;2575
coconut;2564
sesame;2481
cheddar;2396
yolk;2309
shrimp;2250
fillet;2082
steak;2043
lettuce;1922
strawberry;1901
spinach;1889
tortilla;1786
buttermilk;1778
yogurt;1724
mozzarella;1716
rib;1715
turkey;1682
sausage;1639
cucumber;1618
pea;1599
cake;1571
avocado;1570
pineapple;1563
coarse;1560
raspberry;1526
pecan;1497
walnut;1470
wheat;1425
cranberry;1387
chipotle;1357
stemmed;1357
banana;1348
cabbage;1330
fish;1323
pasta;1314
grape;1310
dough;1288
food;1236
zucchini;1214
chilled;1198
map

## Add specific index column

In [17]:
df_82k['id'] = df_82k.index.values
df_82k.head(5)

cooking_method  \
0  ['Set the racks in the middle and upper thirds...   
1  ['Place the eggs in the air fryer basket and c...   
2  ['Air Fryer directions:', 'Preheat air fryer t...   
3  ['Preheat the oven to 400F.', 'Pour 2 tablespo...   
4  ['Cook potatoes in a large pot of salted water...   

                                               image  \
0  https://www.skinnytaste.com/wp-content/uploads...   
1  https://www.skinnytaste.com/wp-content/uploads...   
2  https://www.skinnytaste.com/wp-content/uploads...   
3  https://www.skinnytaste.com/wp-content/uploads...   
4  https://www.skinnytaste.com/wp-content/uploads...   

                                         ingredients  \
0  ['1 tablespoons extra virgin olive oil', '1 cu...   
1  ['4 large eggs', 'Salt (black pepper, everythi...   
2  ['olive oil spray', '4 about 5 ounce each salm...   
3  ['1/2 cup freshly grated Parmesan (not pre-gra...   
4  ['3 1/2 pounds new potatoes (about 10 peeled a...   

                       recipe_name  \
0           Mediterranean Sea Bass   
1       Air Fryer Hard Boiled Eggs   
2  Air Fryer Basil-Parmesan Salmon   
3       Everything Parmesan Crisps   
4      Potato and Green Bean Salad   

                                                tags  \
0  Dairy Free, Gluten Free, Keto Recipes, Kid Fri...   
1  Air Fryer Recipes, Dairy Free, Gluten Free, Ke...   
2  Air Fryer Recipes, Gluten Free, Keto Recipes, ...   
3  Gluten Free, Keto Recipes, Kid Friendly, Low C...   
4  Dairy Free, Gluten Free, Kid Friendly, Vegetar...   

                                  parsed_ingredients vegetarian  \
0  onion garlic tomato their wine kalamata chilea...         no   
1                                          egg bagel         no   
2     salmon fillet lemon mayonnaise parmesan cheese         no   
3  parmesan pre using hole sesame onion garlic poppy        yes   
4                          potato bean bean scallion        yes   

  no_of_servings  id  
0              4   0  
1              4   1  
2              4   2  
3              4   3  
4             12   4

## Add number of ingredients as specific columns

In [18]:
def count_n_ingredients(ingreds):
    ingreds_list = ingreds.split()
    n_ingredients = len(ingreds_list)
    return n_ingredients

In [19]:
df_82k['n_ingredients'] = [count_n_ingredients(x) for x in df_82k['parsed_ingredients']]
df_82k.head(5)

cooking_method  \
0  ['Set the racks in the middle and upper thirds...   
1  ['Place the eggs in the air fryer basket and c...   
2  ['Air Fryer directions:', 'Preheat air fryer t...   
3  ['Preheat the oven to 400F.', 'Pour 2 tablespo...   
4  ['Cook potatoes in a large pot of salted water...   

                                               image  \
0  https://www.skinnytaste.com/wp-content/uploads...   
1  https://www.skinnytaste.com/wp-content/uploads...   
2  https://www.skinnytaste.com/wp-content/uploads...   
3  https://www.skinnytaste.com/wp-content/uploads...   
4  https://www.skinnytaste.com/wp-content/uploads...   

                                         ingredients  \
0  ['1 tablespoons extra virgin olive oil', '1 cu...   
1  ['4 large eggs', 'Salt (black pepper, everythi...   
2  ['olive oil spray', '4 about 5 ounce each salm...   
3  ['1/2 cup freshly grated Parmesan (not pre-gra...   
4  ['3 1/2 pounds new potatoes (about 10 peeled a...   

                       recipe_name  \
0           Mediterranean Sea Bass   
1       Air Fryer Hard Boiled Eggs   
2  Air Fryer Basil-Parmesan Salmon   
3       Everything Parmesan Crisps   
4      Potato and Green Bean Salad   

                                                tags  \
0  Dairy Free, Gluten Free, Keto Recipes, Kid Fri...   
1  Air Fryer Recipes, Dairy Free, Gluten Free, Ke...   
2  Air Fryer Recipes, Gluten Free, Keto Recipes, ...   
3  Gluten Free, Keto Recipes, Kid Friendly, Low C...   
4  Dairy Free, Gluten Free, Kid Friendly, Vegetar...   

                                  parsed_ingredients vegetarian  \
0  onion garlic tomato their wine kalamata chilea...         no   
1                                          egg bagel         no   
2     salmon fillet lemon mayonnaise parmesan cheese         no   
3  parmesan pre using hole sesame onion garlic poppy        yes   
4                          potato bean bean scallion        yes   

  no_of_servings  id  n_ingredients  
0              4   0             15  
1              4   1              2  
2              4   2              6  
3              4   3              8  
4             12   4              4

## Saving the Dataset into CSV file

In [20]:
df_82k.to_csv("../data/parsed_recipes.csv", sep=",", index_label = id)

## Save to .pkl

In [21]:
df_82k.to_pickle("../data/parsed_recipes.pkl")

## Load dataframe

In [22]:
df = pd.read_pickle("../data/parsed_recipes.pkl")
df.head(3)

cooking_method  \
0  ['Set the racks in the middle and upper thirds...   
1  ['Place the eggs in the air fryer basket and c...   
2  ['Air Fryer directions:', 'Preheat air fryer t...   

                                               image  \
0  https://www.skinnytaste.com/wp-content/uploads...   
1  https://www.skinnytaste.com/wp-content/uploads...   
2  https://www.skinnytaste.com/wp-content/uploads...   

                                         ingredients  \
0  ['1 tablespoons extra virgin olive oil', '1 cu...   
1  ['4 large eggs', 'Salt (black pepper, everythi...   
2  ['olive oil spray', '4 about 5 ounce each salm...   

                       recipe_name  \
0           Mediterranean Sea Bass   
1       Air Fryer Hard Boiled Eggs   
2  Air Fryer Basil-Parmesan Salmon   

                                                tags  \
0  Dairy Free, Gluten Free, Keto Recipes, Kid Fri...   
1  Air Fryer Recipes, Dairy Free, Gluten Free, Ke...   
2  Air Fryer Recipes, Gluten Free, Keto Recipes, ...   

                                  parsed_ingredients vegetarian  \
0  onion garlic tomato their wine kalamata chilea...         no   
1                                          egg bagel         no   
2     salmon fillet lemon mayonnaise parmesan cheese         no   

  no_of_servings  id  n_ingredients  
0              4   0             15  
1              4   1              2  
2              4   2              6

## Add dataframe to postgresql database

### Reduce number of recipes
(Free version of Heroku database does not support more than 10000 rows)

In [35]:
df = pd.read_pickle("../data/parsed_recipes.pkl")
print("Number of recipes in preprocessed dataframe: ", len(df))
df.head(3)

Number of recipes in preprocessed dataframe:  55023


cooking_method  \
0  ['Set the racks in the middle and upper thirds...   
1  ['Place the eggs in the air fryer basket and c...   
2  ['Air Fryer directions:', 'Preheat air fryer t...   

                                               image  \
0  https://www.skinnytaste.com/wp-content/uploads...   
1  https://www.skinnytaste.com/wp-content/uploads...   
2  https://www.skinnytaste.com/wp-content/uploads...   

                                         ingredients  \
0  ['1 tablespoons extra virgin olive oil', '1 cu...   
1  ['4 large eggs', 'Salt (black pepper, everythi...   
2  ['olive oil spray', '4 about 5 ounce each salm...   

                       recipe_name  \
0           Mediterranean Sea Bass   
1       Air Fryer Hard Boiled Eggs   
2  Air Fryer Basil-Parmesan Salmon   

                                                tags  \
0  Dairy Free, Gluten Free, Keto Recipes, Kid Fri...   
1  Air Fryer Recipes, Dairy Free, Gluten Free, Ke...   
2  Air Fryer Recipes, Gluten Free, Keto Recipes, ...   

                                  parsed_ingredients vegetarian  \
0  onion garlic tomato their wine kalamata chilea...         no   
1                                          egg bagel         no   
2     salmon fillet lemon mayonnaise parmesan cheese         no   

  no_of_servings  id  n_ingredients  
0              4   0             15  
1              4   1              2  
2              4   2              6

In [39]:
database = df.sample(7000)
database.index = np.arange(len(database))
database['id'] = np.arange(len(database))
database.head(5)

cooking_method  \
0  ['Directions', 'Place all ingredients in a ble...   
1  ['Preheat the oven to 325 degrees. Butter a 9i...   
2  ['Preheat oven on to 375F.', 'Wrap a piece of ...   
3  ['Cut potatoes in half and place in a pot. Cov...   
4  ['Combine wine, Truvia&reg; Spiced Simple Syru...   

                                               image  \
0  https://food.fnr.sndimg.com/content/dam/images...   
1  https://food.fnr.sndimg.com/content/dam/images...   
2  https://www.skinnytaste.com/wp-content/uploads...   
3  https://food.fnr.sndimg.com/content/dam/images...   
4  https://food.fnr.sndimg.com/content/dam/images...   

                                         ingredients  \
0  ['1/4 cup walnut oil', '1/2 teaspoon minced ga...   
1  ['1/2 cup slivered almonds, toasted (see Note)...   
2  ['1.25 lbs 3 boneless skinless chicken breast,...   
3  ['2 to 2 1/4 pounds baby Yukon gold potatoes',...   
4  ['1 bottle (750 ml) red wine', '3/4 cup Truvia...   

                   recipe_name  \
0                       Walnut   
1            Citrus Cheesecake   
2  Bacon Wrapped Chicken Bites   
3      Cheesy Smashed Potatoes   
4                  Red Sangria   

                                                tags  \
0  Salad Dressing Recipes,Nut Recipes,Shellfish R...   
1  Easy Dessert Recipes,Dessert,Easy,Easy Baking,...   
2  Dairy Free, Gluten Free, Keto Recipes, Kid Fri...   
3  Easy Side Dish Recipes,Easy,Side Dish,Thanksgi...   
4                                 Fruit,Wine Recipes   

                                  parsed_ingredients vegetarian  \
0  walnut garlic rich chicken shellfish stock win...         no   
1  slivered cream cheese allpurpose flour sugar l...         no   
2                               chicken breast bacon         no   
3      yukon gold potato cream cheddar chive snipped        yes   
4  wine truvia spiced simple syrup sparkling oran...        yes   

  no_of_servings  id  n_ingredients  
0              0   0             29  
1              8   1             13  
2             10   2              3  
3              4   3              7  
4             10   4             43

### Add table to local postgresql database

In [40]:
from sqlalchemy import create_engine

In [41]:
#engine = create_engine('postgresql://postgres:MySQLPW@localhost:5432/WhatToCook')
engine = create_engine('postgresql://postgres:newPassword@localhost:5432/RecipeRecommendations')

In [43]:
database.to_sql('recipes', engine)

In [44]:
len(database)

7000

### Add table to remote heroku postgresql database

In [46]:
engine = create_engine('<INSERT POSTGRESQL URI HERE>')

In [47]:
database.to_sql('recipes', engine)